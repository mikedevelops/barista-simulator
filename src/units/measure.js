import { Map } from 'immutable';
import { createFilteredCoffeeSubstance, createWaterSubstance } from '../substances/substances'

/**
 * Create measure
 * @param amount
 * @param unit
 * @param substance
 */
const createMeasure = (amount, unit, substance) =>
    Map({ amount, unit, substance });

/**
 * Millilitre name
 * @type {string}
 */
const MEASURE_ML = 'MEASURE_ML';

/**
 * Create Millilitre measure
 * @param amount
 * @param substance
 */
export const createMillilitresMeasure = (amount, substance = null) =>
    createMeasure(amount, MEASURE_ML, substance);

/**
 * Create a Millilitre Water measure
 * @param amount
 */
export const createMillilitresWaterMeasure = (amount) =>
    createMillilitresMeasure(amount, createWaterSubstance());

/**
 * Create a Millilitre Filtered Coffee measure
 * @param amount
 */
export const createMillilitreFilteredCoffeeMeasure = (amount) =>
    createMillilitresMeasure(amount, createFilteredCoffeeSubstance());
