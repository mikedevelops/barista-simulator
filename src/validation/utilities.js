import { Map } from 'immutable';

/**
 * Starting point for all validation
 * @type {number}
 */
export const VALIDATION_SCORE_BASE = 10;

/**
 * Create a new Validation score
 */
export const createValidationScore = () =>
    VALIDATION_SCORE_BASE;

/**
 * Subtract from validation score
 * @param validationScore
 * @param weight
 */
export const deductValidationScore = (validationScore, weight) =>
    validationScore - weight;

/**
 * Create Validation Result
 * @param score
 * @param messages
 */
export const createValidationResult = (score, messages) =>
    Map({ score, messages });
