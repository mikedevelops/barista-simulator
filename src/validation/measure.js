import { createValidationResult, createValidationScore, deductValidationScore } from './utilities'
import { List, Map } from 'immutable';

/**
 * Measure validation weight map
 * @type {*}
 */
export const measureValidationWeights = {
    AMOUNT: (expected, received) => {
        let diff = expected - received;

        if (diff < 0) {
            diff = diff * -1;
        }

        if (diff > 50) {
            return 3;
        }

        if (diff > 25) {
            return 2;
        }

        if (diff > 10) {
            return 1;
        }
    },
    SUBSTANCE: 3
};

/**
 * Validate a full measure
 *
 * @param recipeMeasure
 * @param orderMeasure
 * @param weightMap
 */
export const validateMeasure = (
    recipeMeasure,
    orderMeasure,
    weightMap = measureValidationWeights
) => {
    let validationMessages = List();
    let validationScore = createValidationScore();

    try {
        validateMeasureAmount(recipeMeasure, orderMeasure);
    } catch (err) {
        validationMessages = validationMessages.push(Map({
            message: err.message
        }));
        validationScore = deductValidationScore(
            validationScore,
            weightMap.AMOUNT(
                recipeMeasure.get('amount'),
                orderMeasure.get('amount')));
    }

    try {
        validateMeasureSubstance(recipeMeasure, orderMeasure);
    } catch (err) {
        validationMessages = validationMessages.push(Map({
            message: err.message
        }));
        validationScore = deductValidationScore(
            validationScore,
            weightMap.SUBSTANCE);
    }

    return createValidationResult(validationScore, validationMessages);
};

/**
 * Validate measure amount
 *
 * @param recipeMeasure
 * @param orderMeasure
 * @returns {boolean}
 */
export const validateMeasureAmount  = (recipeMeasure, orderMeasure) => {
    const recipeMeasureAmount = recipeMeasure.get('amount');
    const orderMeasureAmount = orderMeasure.get('amount');

    if (recipeMeasureAmount !== orderMeasureAmount) {
        throw new Error(`"${orderMeasureAmount}" is the wrong amount. ` +
            `Expected "${recipeMeasureAmount}"`);
    }

    return true;
}

/**
 * Validate Measure substance
 *
 * @param recipeMeasure
 * @param orderMeasure
 * @throws
 * @returns {boolean}
 */
export const validateMeasureSubstance  = (recipeMeasure, orderMeasure) => {
    const recipeMeasureSubstance = recipeMeasure.getIn(['substance', 'name']);
    const orderMeasureSubstance = orderMeasure.getIn(['substance', 'name']);

    if (recipeMeasureSubstance !== orderMeasureSubstance) {
        throw new Error(`"${orderMeasureSubstance}" is the wrong substance. ` +
            `Expected "${recipeMeasureSubstance}"`);
    }

    return true;
}
