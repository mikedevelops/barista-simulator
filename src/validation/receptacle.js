import { createValidationResult, createValidationScore, deductValidationScore } from './utilities'
import { List, Map } from 'immutable';

export const receptacleValidationWeights = {
    TYPE: () => -2
};

export const validateReceptacle = (
    recipeReceptacle,
    orderReceptacle,
    weightMap = receptacleValidationWeights
) => {
    let validationMessages = List();
    let validationScore = createValidationScore();

    try {
        validateReceptacleType(recipeReceptacle, orderReceptacle);
    } catch (err) {
        validationMessages = validationMessages.push(Map({
            message: err.message
        }));
        validationScore = deductValidationScore(
            validationScore,
            weightMap.TYPE());
    }

    return createValidationResult(validationScore, validationMessages);
}

/**
 * Validate Receptacle
 *
 * @param recipeReceptacle
 * @param orderReceptacle
 * @throws
 * @returns boolean
 */
export const validateReceptacleType = (recipeReceptacle, orderReceptacle) => {
    if (recipeReceptacle.get('name') !== orderReceptacle.get('name')) {
        throw new Error(`"${orderReceptacle.get('name')}" is the wrong receptacle. ` +
            `Expected a "${recipeReceptacle.get('name')}".`);
    }

    return true;
}
