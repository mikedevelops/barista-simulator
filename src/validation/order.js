import { createValidationScore } from './utilities';
import { validateReceptacle } from './receptacle'
import { validateMeasure } from './measure'

export const validateOrder = (expectedOrder, receivedReceptacle) => {
    // validate receptacle
    let score = validateReceptacle(expectedOrder.get('receptacle'), receivedReceptacle);

    // validate ingredients
    expectedOrder.getIn(['recipe', 'ingredients']).forEach(ingredient => {
        const receivedIngredient = receivedReceptacle.get('contents').find(content =>
                content.getIn(['substance', 'name']) === ingredient.getIn(['substance', 'name']));

        if (receivedIngredient !== undefined) {
            score = validateMeasure(ingredient, receivedIngredient);
        }
    });

    return score;
}
