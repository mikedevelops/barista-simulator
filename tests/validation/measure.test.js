import { createMillilitreFilteredCoffeeMeasure, createMillilitresWaterMeasure } from '../../src/units/measure'
import { validateMeasure, validateMeasureAmount, validateMeasureSubstance } from '../../src/validation/measure'
import { List, Map, is } from 'immutable';

describe('measure validation', () => {
    describe('validateMeasureAmount', () => {
        test('Should throw if the amounts do not match', () => {
            const recipeMeasure = createMillilitresWaterMeasure(200);
            const orderMeasure = createMillilitresWaterMeasure(201);

            expect(() => validateMeasureAmount(recipeMeasure, orderMeasure)).toThrowErrorMatchingSnapshot();
        });

        test('Should return true if the amounts match', () => {
            const recipeMeasure = createMillilitresWaterMeasure(200);
            const orderMeasure = createMillilitresWaterMeasure(200);

            expect(validateMeasureAmount(recipeMeasure, orderMeasure)).toBe(true);
        });
    });

    describe('validateMeaxsureSubstance', () => {
        test('Should throw if the substance names do not match', () => {
            const recipeMeasure = createMillilitreFilteredCoffeeMeasure(200);
            const orderMeasure = createMillilitresWaterMeasure(200);

            expect(() => validateMeasureSubstance(recipeMeasure, orderMeasure)).toThrowErrorMatchingSnapshot();
        });

        test('Should return true if the substances match', () => {
            const recipeMeasure = createMillilitresWaterMeasure(200);
            const orderMeasure = createMillilitresWaterMeasure(200);

            expect(validateMeasureSubstance(recipeMeasure, orderMeasure)).toBe(true);
        });
    });

    describe('validateMeasure', () => {
        let weightMap;

        beforeEach(() => {
            weightMap = {
                AMOUNT: (expected, received) => Math.floor((expected - received) / 10)
            }
        });

        test('should return the base validation score and no messages for identical measures', () => {
            const recipeMeasure = createMillilitresWaterMeasure(100);
            const orderMeasure = createMillilitresWaterMeasure(100);
            const score = Map({
                score: 10,
                messages: List()
            });

            expect(is(validateMeasure(recipeMeasure, orderMeasure, weightMap), score)).toBe(true);
        });

        test('should return a validation score with messages for invalid measures', () => {
            const recipeMeasure = createMillilitresWaterMeasure(150);
            const orderMeasure = createMillilitresWaterMeasure(100);

            expect(validateMeasure(recipeMeasure, orderMeasure, weightMap).get('score')).toBe(5);
            expect(validateMeasure(recipeMeasure, orderMeasure, weightMap).get('messages').size).toBe(1);
        });
    });
});
