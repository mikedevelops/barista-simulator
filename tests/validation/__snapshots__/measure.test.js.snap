// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`measure validation validateMeasureAmount Should throw if the amounts do not match 1`] = `"\\"201\\" is the wrong amount. Expected \\"200\\""`;

exports[`measure validation validateMeaxsureSubstance Should throw if the substance names do not match 1`] = `"\\"SUBSTANCE_WATER\\" is the wrong substance. Expected \\"SUBSTANCE_FILTERED_COFFEE\\""`;
